server/src/main/java/io/deephaven/server/barrage/BarrageStreamGenerator.java

```diff
@@ -298,34 +379,126 @@ public SubView(final BarrageStreamGenerator generator,
-            }
-            offset = 0;
-            for (long ii = 0; ii < numModBatches; ++ii) {
+
+            // batch size is maximum, will write fewer rows when needed. If we are enforcing a message size limit, then
+            // reasonably we will require at least one byte per row and restrict rows accordingly
+            int batchSizeLimit = Math.min(DEFAULT_MESSAGE_SIZE_LIMIT, batchSize());
+
+            // there may be multiple chunk generators and we need to ensure that a single batch rowset does not cross
+            // boundaries and that we honor the batch size requests
+
+            if (numAddRows == 0 && numModRows == 0) {
+                // we still need to send a message containing metadata when there are no rows
                 final InputStream is = generator.getInputStream(
-                        this, offset, offset + batchSize, metadata, generator::appendModColumns);
+                        this, 0, 0, metadata, generator::appendAddColumns);
                 bytesWritten += is.available();
                 visitor.accept(is);
-                offset += batchSize;
-                metadata = null;
+            } else {
+                long offset = 0;
+                long chunkOffset = 0;
+
+                while (offset < numAddRows) {
+                    long rowsRemaining = numAddRows - offset;
+
+                    int batchSize = (int) Math.min(rowsRemaining, batchSizeLimit);
+
+                    // make sure we are not about to cross a chunk boundary
+                    if (chunkOffset + batchSize > TWO_DIMENSIONAL_COLUMN_SOURCE_THRESHOLD) {
+                        batchSize = (int) (TWO_DIMENSIONAL_COLUMN_SOURCE_THRESHOLD - chunkOffset);
+                        chunkOffset = 0;
+                    }
+
+                    try {
+                        final InputStream is = generator.getInputStream(
+                                this, offset, offset + batchSize, metadata, generator::appendAddColumns);
+                        int bytesToWrite = is.available();
+
+                        // treat this as a hard limit, exceeding fails a client or w2w (unless we are sending a single
+                        // row then we must send and let it potentially fail)
+                        if (bytesToWrite > DEFAULT_MESSAGE_SIZE_LIMIT && batchSize > 1) {
+                            // can't write this, so close the input stream and retry
+                            is.close();
+                        } else {
+                            bytesWritten += bytesToWrite;
+
+                            // let's write the data
+                            visitor.accept(is);
+
+                            offset += batchSize;
+                            chunkOffset += batchSize;
+                            metadata = null;
+                        }
+                        // recompute the batch limit for the next message
+                        int bytesPerRow = bytesToWrite / batchSize;
```

Comment:
Maybe add a comment that `getInputStream` might append fewer rows if this batch does not overlap completely with the user's subscription. This is likely ok as it just underestimates bytes per row.

I wonder if it would be worth having `appendAllColumns` try to write the intended range as sliced from `myAddedOffsets`. You could pass a `MutableInt` into `getInputStream` that writes how many rows were actually written -- allowing you to short-cut the case where the "batch" crosses two chunk input stream generators.

Actually, what do you think of that idea? It will eliminate sending a bunch of empty batches to viewports when you have a full subscription and a viewport subscribed to the same table. It is complicated for modifications as the chunk boundaries are not guaranteed to line up. Maybe the compromise for mods is to run through all columns to limit the size of the batch before creating the first column's input stream. It feels like a good compromise.